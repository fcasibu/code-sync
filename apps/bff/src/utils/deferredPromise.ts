export class DeferredPromise<T> {
  public promise: Promise<T>;

  public resolve: (value: T | PromiseLike<T>) => void = () => undefined;
  public reject: (reason?: unknown) => void = () => undefined;

  constructor() {
    this.promise = new Promise((res, rej) => {
      this.resolve = res;
      this.reject = rej;
    });
  }
}
