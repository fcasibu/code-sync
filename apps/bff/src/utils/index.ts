export * from './urlForHTTPServer';
export * from './gqlHelper';
export * from './testHelpers';
export * from './lazy';
export * from './typeGuards';
