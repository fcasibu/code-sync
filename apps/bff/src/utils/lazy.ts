export const lazy = <T>(factory: () => T) => {
  let instance: T;

  return () => {
    if (!instance) {
      instance = factory();
    }

    return instance;
  };
};
