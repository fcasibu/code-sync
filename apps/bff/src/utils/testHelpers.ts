import type { PrismaClient } from '@code-sync/db';
import { prismaClient } from '../config';
import type { Application } from '../server';
import { startServer } from '../server';

type UnMaybe<T> = {
  [P in keyof T]-?: NonNullable<T[P]>;
};

export const testHelper = () => {
  let _app: Application | undefined;
  let _prismaClient: PrismaClient | undefined;

  const setup = async () => {
    _prismaClient = prismaClient();
    _app = await startServer({ port: 0, prismaClient: _prismaClient });
  };

  const teardown = async () => {
    await _app?.apolloServer?.stop();
    await _prismaClient?.$disconnect();
    _app?.expressServer?.close();
  };

  return {
    setup,
    teardown,
    get app() {
      if (!_app)
        throw new Error(
          '`setup` in the testHelper was not invoked. Invoke the `setup` of `testHelper` in either `beforeEach` or `beforeAll`',
        );

      return _app as UnMaybe<Application>;
    },
    get prismaClient() {
      if (!_prismaClient)
        throw new Error(
          '`setup` in the testHelper was not invoked. Invoke the `setup` of `testHelper` in either `beforeEach` or `beforeAll`',
        );

      return _prismaClient;
    },
  };
};
