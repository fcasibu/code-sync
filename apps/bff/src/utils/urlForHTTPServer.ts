import type { Server } from 'node:http';
import type { AddressInfo } from 'node:net';
import { format } from 'node:url';

// https://github.com/apollographql/apollo-server/blob/main/packages/server/src/utils/urlForHttpServer.ts
export function urlForHTTPServer(expressServer: Server) {
  const { address, port } = expressServer.address() as AddressInfo;

  // Convert IPs which mean "any address" (IPv4 or IPv6) into localhost
  // corresponding loopback ip. Note that the url field we're setting is
  // primarily for consumption by our test suite. If this heuristic is wrong for
  // your use case, explicitly specify a frontend host (in the `host` option
  // when listening).
  const hostname = address === '' || address === '::' ? 'localhost' : address;

  return format({
    protocol: 'http',
    hostname,
    port,
    pathname: '/',
  });
}
