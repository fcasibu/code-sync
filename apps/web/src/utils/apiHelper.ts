import { type NextRequest, NextResponse } from 'next/server';
import { INTERNAL_HEADER_API_KEY } from '@/constants';
import { env } from '@/env';

export const isAuthorized = (req: NextRequest) =>
  req.headers.get(INTERNAL_HEADER_API_KEY) ===
  env.INTERNAL_HEADER_API_KEY_VALUE;

export const sendUnauthorizedResponse = () =>
  NextResponse.json({ error: 'Unauthorized' }, { status: 401 });
