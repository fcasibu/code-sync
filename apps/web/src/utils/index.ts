export * from './gqlHelper';
export * from './runtime';
export * from './apiHelper';
