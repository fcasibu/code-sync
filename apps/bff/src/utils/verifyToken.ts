import { promisify } from 'node:util';
import { strict as assert } from 'assert';
import type { JwtPayload } from 'jsonwebtoken';
import jwt from 'jsonwebtoken';
import { JwksClient } from 'jwks-rsa';
import { logger } from '@code-sync/logger';
import { DeferredPromise } from './deferredPromise';

export const verifyToken = (token: string) => {
  const client = new JwksClient({
    jwksUri: `https://${process.env.AUTH0_DOMAIN}/.well-known/jwks.json`,
  });

  const deferredPromise = new DeferredPromise<jwt.JwtPayload>();

  const getSigningKey = promisify(client.getSigningKey);

  const getJwksClientKey = async (
    header: jwt.JwtHeader,
    callback: jwt.SigningKeyCallback,
  ) => {
    const key = await getSigningKey(header.kid);
    callback(null, key?.getPublicKey());
  };

  jwt.verify(
    token,
    getJwksClientKey,
    getJwtVerifyOptions(),
    (error, decoded) => {
      if (error) {
        logger.error(error, 'Was unable to verify token');
        deferredPromise.resolve({});
        return;
      }

      deferredPromise.resolve(decoded as JwtPayload);
    },
  );

  return deferredPromise.promise;
};

const getJwtVerifyOptions = (): jwt.VerifyOptions => {
  const { AUTH0_AUDIENCE, AUTH0_DOMAIN } = process.env;

  assert(AUTH0_AUDIENCE, 'AUTH0_AUDIENCE is required environment variable.');
  assert(AUTH0_DOMAIN, 'AUTH0_DOMAIN is required environment variable.');

  return {
    audience: AUTH0_AUDIENCE,
    issuer: `https://${AUTH0_DOMAIN}/`,
    algorithms: ['RS256'],
  };
};
