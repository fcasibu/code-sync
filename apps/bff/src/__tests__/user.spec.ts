import { faker } from '@faker-js/faker';
import requestGQL from 'supertest-graphql';
import { afterAll, beforeAll, describe, expect, it } from 'vitest';
import type { ResultOf, VariablesOf } from '../utils/gqlHelper';
import { graphql } from '../utils/gqlHelper';
import { testHelper } from '../utils/testHelpers';

const helper = testHelper();

beforeAll(helper.setup);
afterAll(helper.teardown);

describe('resolver#User', () => {
  describe('users', () => {
    it('should return a list of users', async () => {
      const query = graphql(`
        query UserQuery {
          users {
            id
            displayName
            email
            authId
            authProvider
            profilePicture
            rooms {
              id
              name
              owner {
                id
                rooms {
                  __typename
                }
                authProvider
                authId
                email
                displayName
                profilePicture
              }
            }
          }
        }
      `);

      const response = await requestGQL<ResultOf<typeof query>>(
        helper.app.expressServer,
      )
        .query(query)
        .expectNoErrors();

      expect(response.data?.users).not.toHaveLength(0);
      expect(response.data).toMatchSnapshot();
    });
  });

  describe('user', () => {
    it('should return a user through their ID', async () => {
      const query = graphql(`
        query UserQuery($input: UserByIdInput!) {
          user(input: $input) {
            id
            displayName
            email
            authId
            authProvider
            profilePicture
            rooms {
              id
              name
              owner {
                id
                rooms {
                  __typename
                }
                authProvider
                authId
                email
                displayName
                profilePicture
              }
            }
          }
        }
      `);

      const userId = 'user-0';
      const response = await requestGQL<
        ResultOf<typeof query>,
        VariablesOf<typeof query>
      >(helper.app.expressServer)
        .query(query)
        .variables({ input: { id: userId } })
        .expectNoErrors();

      expect(response.data?.user?.id).toBe(userId);
      expect(response.data).toMatchSnapshot();
    });
  });

  describe('createUser', () => {
    it('should create a user and return it', async () => {
      const mutation = graphql(`
        mutation UserMutation($input: CreateUserInput!) {
          createUser(input: $input) {
            id
            displayName
            email
            authId
            authProvider
            profilePicture
            rooms {
              id
              name
              owner {
                id
                rooms {
                  __typename
                }
                authProvider
                authId
                email
                displayName
                profilePicture
              }
            }
          }
        }
      `);

      const variables: VariablesOf<typeof mutation> = {
        input: {
          email: faker.internet.email(),
          profilePicture: faker.image.avatar(),
          authId: faker.string.nanoid(),
          authProvider: 'github',
          displayName: faker.internet.displayName(),
        },
      };
      const response = await requestGQL<
        ResultOf<typeof mutation>,
        VariablesOf<typeof mutation>
      >(helper.app.expressServer)
        .mutate(mutation)
        .variables(variables)
        .expectNoErrors();

      expect(response.data?.createUser?.authProvider).toBe(
        variables.input.authProvider,
      );
      expect(response.data?.createUser?.authId).toBe(variables.input.authId);
      expect(response.data?.createUser?.email).toBe(variables.input.email);
      expect(response.data?.createUser?.displayName).toBe(
        variables.input.displayName,
      );
      expect(response.data?.createUser?.profilePicture).toBe(
        variables.input.profilePicture,
      );
    });
  });
});
