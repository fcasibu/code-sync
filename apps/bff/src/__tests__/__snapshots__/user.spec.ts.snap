// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`resolver#User > user > should return a user through their ID 1`] = `
{
  "user": {
    "authId": "NwzV451PMRYNsM7D8Wwpg",
    "authProvider": "github",
    "displayName": "Katharina",
    "email": "Katharina_Sporer@yahoo.com",
    "id": "user-0",
    "profilePicture": "https://via.placeholder.com/150x150/bc9de8/96d1f5.jpeg?text=Katharina",
    "rooms": null,
  },
}
`;

exports[`resolver#User > users > should return a list of users 1`] = `
{
  "users": [
    {
      "authId": "NwzV451PMRYNsM7D8Wwpg",
      "authProvider": "github",
      "displayName": "Katharina",
      "email": "Katharina_Sporer@yahoo.com",
      "id": "user-0",
      "profilePicture": "https://via.placeholder.com/150x150/bc9de8/96d1f5.jpeg?text=Katharina",
      "rooms": null,
    },
    {
      "authId": "jwrexUsqWMIN5wRPPg3FA",
      "authProvider": "github",
      "displayName": "Hayley",
      "email": "Hayley.Watsica30@hotmail.com",
      "id": "user-1",
      "profilePicture": "https://via.placeholder.com/150x150/f35284/87e8eb.webp?text=Hayley",
      "rooms": null,
    },
    {
      "authId": "Cw8vbMQuQ4-4q57Ae6dlt",
      "authProvider": "github",
      "displayName": "Hudson",
      "email": "Hudson_Waters@gmail.com",
      "id": "user-2",
      "profilePicture": "https://via.placeholder.com/150x150/9a9773/4fd26a.jpeg?text=Hudson",
      "rooms": null,
    },
    {
      "authId": "9sdluULYoyiAhsKOfTI1_",
      "authProvider": "github",
      "displayName": "Kathryne",
      "email": "Kathryne.Spencer51@gmail.com",
      "id": "user-3",
      "profilePicture": "https://via.placeholder.com/150x150/cd9d02/045129.png?text=Kathryne",
      "rooms": null,
    },
    {
      "authId": "0zm-oOBgDO12Wrr4aRQVR",
      "authProvider": "github",
      "displayName": "Vito",
      "email": "Vito71@hotmail.com",
      "id": "user-4",
      "profilePicture": "https://via.placeholder.com/150x150/dfd43f/40f5d9.png?text=Vito",
      "rooms": null,
    },
    {
      "authId": "EOfL6c7SwuJ05PPTXy5Zt",
      "authProvider": "github",
      "displayName": "Damaris",
      "email": "Damaris.Sipes91@gmail.com",
      "id": "user-5",
      "profilePicture": "https://via.placeholder.com/150x150/fa1d8a/7ba6f4.webp?text=Damaris",
      "rooms": null,
    },
    {
      "authId": "Koe9I-tGNsF9mCrz5ddPs",
      "authProvider": "github",
      "displayName": "Sister",
      "email": "Sister4@hotmail.com",
      "id": "user-6",
      "profilePicture": "https://via.placeholder.com/150x150/82fd2b/d9b7ad.webp?text=Sister",
      "rooms": null,
    },
    {
      "authId": "GNn4hUdsV1BWeyU8wKFtc",
      "authProvider": "github",
      "displayName": "Arden",
      "email": "Arden_Hamill@gmail.com",
      "id": "user-7",
      "profilePicture": "https://via.placeholder.com/150x150/faf6b5/b11950.jpg?text=Arden",
      "rooms": null,
    },
    {
      "authId": "9HoO1lt-3gi-ibrwWvF_k",
      "authProvider": "github",
      "displayName": "Schuyler",
      "email": "Schuyler_Fisher-Strosin@gmail.com",
      "id": "user-8",
      "profilePicture": "https://via.placeholder.com/150x150/b590d9/a88bda.webp?text=Schuyler",
      "rooms": null,
    },
    {
      "authId": "aSW2HzTFxHnHPsWNYawWE",
      "authProvider": "github",
      "displayName": "Adell",
      "email": "Adell.Kutch@yahoo.com",
      "id": "user-9",
      "profilePicture": "https://via.placeholder.com/150x150/26304d/935318.jpeg?text=Adell",
      "rooms": null,
    },
    {
      "authId": "GX0vOBIiUIx8nFrqCvEBN",
      "authProvider": "github",
      "displayName": "Olen",
      "email": "Olen.Kunde@gmail.com",
      "id": "user-10",
      "profilePicture": "https://via.placeholder.com/150x150/acac40/4fa535.gif?text=Olen",
      "rooms": null,
    },
    {
      "authId": "J6TRaD1__9LOlj65IB3to",
      "authProvider": "github",
      "displayName": "Chelsie",
      "email": "Chelsie.OReilly@gmail.com",
      "id": "user-11",
      "profilePicture": "https://via.placeholder.com/150x150/fe92a4/0dace2.webp?text=Chelsie",
      "rooms": null,
    },
    {
      "authId": "FYtUb9UGXsyNWdHucD83a",
      "authProvider": "github",
      "displayName": "Cara",
      "email": "Cara.Swift53@yahoo.com",
      "id": "user-12",
      "profilePicture": "https://via.placeholder.com/150x150/f07e0d/601ff3.jpeg?text=Cara",
      "rooms": null,
    },
    {
      "authId": "FUIturwZCbIWxtCGh3Mwc",
      "authProvider": "github",
      "displayName": "Orion",
      "email": "Orion96@yahoo.com",
      "id": "user-13",
      "profilePicture": "https://via.placeholder.com/150x150/3fa9ad/dbbe8f.jpeg?text=Orion",
      "rooms": null,
    },
    {
      "authId": "xfX9dilZeY6aNVx57gIaY",
      "authProvider": "github",
      "displayName": "Jordan",
      "email": "Jordan0@hotmail.com",
      "id": "user-14",
      "profilePicture": "https://via.placeholder.com/150x150/364f4b/778bb4.png?text=Jordan",
      "rooms": null,
    },
    {
      "authId": "2vyZqPHM7m-g5XgetlmJl",
      "authProvider": "github",
      "displayName": "Rosie",
      "email": "Rosie.McLaughlin36@hotmail.com",
      "id": "user-15",
      "profilePicture": "https://via.placeholder.com/150x150/c6eb06/c6f245.webp?text=Rosie",
      "rooms": null,
    },
    {
      "authId": "y4I9fiHsdr7Uztj2KxLN4",
      "authProvider": "github",
      "displayName": "Damien",
      "email": "Damien.Bednar@gmail.com",
      "id": "user-16",
      "profilePicture": "https://via.placeholder.com/150x150/137f43/fe3d1c.webp?text=Damien",
      "rooms": null,
    },
    {
      "authId": "Kz2dug7in0tw2kpiFwLav",
      "authProvider": "github",
      "displayName": "Eleanora",
      "email": "Eleanora_Sporer6@gmail.com",
      "id": "user-17",
      "profilePicture": "https://via.placeholder.com/150x150/bb9e2e/4cd0c7.gif?text=Eleanora",
      "rooms": null,
    },
    {
      "authId": "kyAhDchCoyUwWJgqYbxuM",
      "authProvider": "github",
      "displayName": "Mable",
      "email": "Mable18@gmail.com",
      "id": "user-18",
      "profilePicture": "https://via.placeholder.com/150x150/dc4a0e/9fc5c0.webp?text=Mable",
      "rooms": null,
    },
    {
      "authId": "Qx3IgBRtbwDyBH8YkUBMO",
      "authProvider": "github",
      "displayName": "Axel",
      "email": "Axel_Jerde13@yahoo.com",
      "id": "user-19",
      "profilePicture": "https://via.placeholder.com/150x150/57e10e/2b0ffd.png?text=Axel",
      "rooms": null,
    },
  ],
}
`;
